File: quicksort.cpp
#include <iostream>
#include <stdlib.h>



int quickSort(int *array, int start, int end);
int partition(int *array, int start, int end);
void swap(int *array, int i, int j);

int quickSort(int *array, int start, int end) {
    if (array == NULL || start < 0 || end < start) {
        return -1;
    }

    int smallIndex = partition(array, start, end);
    if (smallIndex > start) {
        quickSort(array, start, smallIndex-1);
    }

    if (smallIndex < end) {
        quickSort(array, smallIndex+1, end);
    }

    return 0;
}

int partition(int *array, int start, int end) {
    int privot = start + rand()%(end - start + 1);
    swap(array, privot, end);
    int smallIndex = -1;
    for (int i = start; i < end; i++) {
        if (array[i] <= array[end]) {
            smallIndex++;
            if (i > smallIndex) {
                swap(array, smallIndex, i);
            }
        }
    }

    return smallIndex;
}

void swap(int *array, int i, int j) {
    int temp = array[i];
    array[i] = array[j];
    array[j] = temp;
}

int main() {
    int arr[6] = {4, 3, 2, 6, 5, 1};
    quickSort(arr, 0, 5);
    return 0;
}

File: quickSort.go
package main

import (
	"fmt"
	"math/rand"
)

func QuickSort(data []int, start, end int) {
	if data == nil || start < 0 || end < 0 || start > end {
		return
	}

	index := partition(data, start, end)

	if index > start {
		QuickSort(data, start, index-1)
	}

	if index < end {
		QuickSort(data, index+1, end)
	}
}

func partition(data []int, start, end int) int {
	index := start + rand.Intn(end-start+1)
	data[index], data[end] = data[end], data[index]

	small := start - 1

	for i := start; i <= end; i++ {
		if data[i] < data[end] {
			small = small + 1
			if small != i {
				data[small], data[i] = data[i], data[small]
			}
		}
	}
	small = small + 1
	data[end], data[small] = data[small], data[end]

	return small
}

func main() {
	test := []int{1, 3, 5, 2, 6, 4}

	QuickSort(test, 0, len(test)-1)
	for _, elm := range test {
		fmt.Printf("%d ", elm)
	}
	fmt.Println()
}
